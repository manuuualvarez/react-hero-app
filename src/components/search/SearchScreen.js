import React, { useMemo } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm'
// Helpers
import { getHeroByName } from '../../helpers/getHeroeByName';
// Components
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();
    const { q = ''} = queryString.parse(location.search);
    
    const [formValues, handleInputChange] = useForm({
        searchText:q
    });

    const { searchText } = formValues;
    const heroFiltered = useMemo( () => (getHeroByName(q)), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${searchText}`);
    }

    return (
        <>
            <h1>Search Screen</h1>
            <hr></hr>

            <div className="row">
                <div className="col-5">
                    <h4>Search</h4>
                    <hr></hr>
                    
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            placeholder="Search an Hero"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value= { searchText }
                            onChange={handleInputChange}
                        >
                        </input>

                        <button type="submit" className="btn btn-primary mt-3 col-12"> Search...</button>
                    </form>

                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr></hr>

                    {
                        (q === '')
                        ?  <div className="alert alert-warning animate__animated animate__bounce">Please search some</div>   
                        : (heroFiltered.length === 0) && 
                        <div className="alert alert-danger animate__animated animate__bounce">
                            No heros with this filter with: {q}
                        </div>
                    }

                    { heroFiltered.map ( (hero) => (
                       <HeroCard 
                            key={hero.id}
                            {...hero} />
                        ))    
                    }
                </div>
            </div>
        </>
    )
}
