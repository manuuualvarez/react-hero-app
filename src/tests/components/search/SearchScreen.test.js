import { mount } from "enzyme"
import { SearchScreen } from '../../../components/search/SearchScreen';
import { MemoryRouter } from 'react-router-dom';


const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));


describe('Test Search Screen', () => {

    test('should show success without parameters', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-warning').text().trim()).toEqual('Please search some'); 
    });

    test('should show success with batman parameter', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });

    test('should show danger alert when the result is empty', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=pokemon'] }>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').text().trim()).toEqual('No heros with this filter with: pokemon'); 
    });

    test('should call the navigate function with the queryString', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', 
        { target: 
            { name: 'searchText',
            value: 'pokemon' } 
        });

        wrapper.find('form').prop('onSubmit')({ preventDefault: () => {} });
        expect(mockNavigate).toHaveBeenCalledWith('?q=pokemon');
        
    })
    


    
    
})
