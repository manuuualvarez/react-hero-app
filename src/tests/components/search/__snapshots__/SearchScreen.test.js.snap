// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Search Screen should show danger alert when the result is empty 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search an Hero"
          type="text"
          value="pokemon"
        />
        <button
          className="btn btn-primary mt-3 col-12"
          type="submit"
        >
           Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger animate__animated animate__bounce"
      >
        No heros with this filter with: 
        pokemon
      </div>
    </div>
  </div>,
]
`;

exports[`Test Search Screen should show success 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search an Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary mt-3 col-12"
          type="submit"
        >
           Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-warning animate__animated animate__bounce"
      >
        Please search some
      </div>
    </div>
  </div>,
]
`;

exports[`Test Search Screen should show success with batman parameter 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search an Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary mt-3 col-12"
          type="submit"
        >
           Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                   
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                     
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                   MÃ¡s Info ... 
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Test Search Screen should show success without parameters 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search an Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary mt-3 col-12"
          type="submit"
        >
           Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-warning animate__animated animate__bounce"
      >
        Please search some
      </div>
    </div>
  </div>,
]
`;
